behaviors/services/trie.md

```diff
@@ -0,0 +1,97 @@
+        * Missing Key indication - if the proof ends with a branch node and the digit in the requested path does not reference any child node, or, if the final node indicates a divergence from the requested key in a way that negates any possibility that the requested key may be present in the trie.
+        * Present Key with value hash indication - if the proof ends with a lead node and the assembled key prefix is identical to the requested path
+* Prefixes are represented as hexadecimal digits in base64 encoding.
+* Zero values are never represented in the trie because they are not part of the state snapthot
+* Each node in a Merkle Trie is addressable via it’s hash-code (cryptographic digest) in a __content-addressed__ scheme - addresses of nodes are derived from their content as a cryptographic hashcode digest.  
+* Each block height represents a (possibly) distinct state in it’s virtual chain. For any block height  with a full Merkle trie and a Merkle Trie root corresponding to the matching state image.
+
+## Hash Function 
+TBD
+## Serialization 
+TBD - decide between
+ * Option 1 - custom - specify optimized serialization… ?
+ * Option 2 - Membuffers - possible issues with size
+ * Option 3 - RLP - for code simplicity and gas savings on EVM
+ 
+## Trie Key limitations
+### Uniform Key Length 
+All trie keys are of a uniform length. This means branch nodes and value nodes are mutually exclusive as there may not be two keys where one is contained in the other.
+   * Simplify trie structure by forcing values to be only in leaf nodes
+   * This means clients interpreting proofs must be aware of the uniform key length and enforce this knowledge when evaluating proofs
+   * This requirement depends on our state addressing scheme. See state storage requirements
+## Node types and structure 
+### Common properties
+All nodes have a prefix part. A Prefix part is the current node’s contribution to the key prefix represented by a path in the trie 
+* For child nodes (of any type - branch or leaf) if the parent node had an “Even” parity bit, we encode the prefix part with an offset of 4 bits (Even and not Odd because the parent’s branch digit reverses the parity once it is appended to the parent prefix part).
+
+### Branch nodes
+In addition to the prefix part attribute, a branch node has up to 16 children.
+Each child reference corresponds to a possible prefix extension. Prefixes are represented as hexadecimal digits in base64 encoding. This is why any possible extension of the current path's prefix may be any one of 16 
+* Parity bit always refers to the number of digits from the beginning of the prefix up to and including the current node’s prefix part. 
+
+### Leaf nodes 
+Leaf nodes represent values or hashes of values associated with the state key which is represetned by the assembled prefix parts of the entire trie path leading to the leaf node from the root of the trie. 
```

Comment:
is it allowed for a leaf have a prefix

Reply:
this is specified in the `## Common properties` section. what do you mean by this comment?